<a href="https://colab.research.google.com/github/PierfrancescoElia/AICyberLabs/blob/main/Lab02/lab02.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Project 2: Model Engineering

| Member | ID |
| -- | -- |
| Alessandro Meneghini| s332228 |
| Pierfrancesco Elia | s331497 |
| Ankesh Porwal | s328746 |


In [2]:
# importing libraries
import pandas as pd
import numpy as np
import json
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

In [11]:
# importing dataset
dataset_url = 'https://raw.githubusercontent.com/PierfrancescoElia/AICyberLabs/refs/heads/main/Lab02/train.json'
train_df = pd.read_json(dataset_url)

dataset_url = 'https://raw.githubusercontent.com/PierfrancescoElia/AICyberLabs/refs/heads/main/Lab02/test.json'
test_df = pd.read_json(dataset_url)

train_df.head()

md5hash  \
0  aaec2a99e23b2478ce5ec7f9b43c840c   
1  14a9db02dd903a76adb275cf45d10413   
2  70f77a07109abab1983d98214f4eb199   
3  fa67eebd3047eb2a4c9379623219d58d   
4  d9209cdfe4fd24c6f2a17acdafa558a5   

                                   api_call_sequence  is_malware  
0  [NtClose, NtOpenKey, NtQueryValueKey, NtClose,...           1  
1  [GetSystemTimeAsFileTime, LdrLoadDll, LdrGetPr...           1  
2  [RegOpenKeyExA, NtOpenKey, NtQueryValueKey, Nt...           1  
3  [RegOpenKeyExA, NtOpenKey, NtQueryValueKey, Nt...           1  
4  [NtClose, NtOpenKey, NtQueryValueKey, NtClose,...           1

Let's join api_call_squence

In [9]:
train_df["api_sequence_str"] = train_df["api_call_sequence"].apply(lambda x: " ".join(x))
test_df["api_sequence_str"] = test_df["api_call_sequence"].apply(lambda x: " ".join(x))

vectorizer = CountVectorizer()
X_train = vectorizer.fit_transform(train_df["api_sequence_str"])
X_test = vectorizer.transform(test_df["api_sequence_str"])  # transform so that out-of-vocaboulary are ignored correctly

y_train = train_df["is_malware"]
y_test = test_df["is_malware"]

how many vocab do i have

In [10]:
print(f"Number of features: {len(vectorizer.get_feature_names_out())}")


Number of features: 253


STEP 4

In [12]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, confusion_matrix

clf = LogisticRegression(max_iter=1000)
clf.fit(X_train, y_train)

y_pred = clf.predict(X_test)


In [17]:
print("Classification Report:\n")
print(classification_report(y_test, y_pred, target_names=["Goodware", "Malware"]))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred))

Classification Report:

              precision    recall  f1-score   support

    Goodware       0.65      0.29      0.40       324
     Malware       0.98      1.00      0.99     12839

    accuracy                           0.98     13163
   macro avg       0.82      0.64      0.69     13163
weighted avg       0.97      0.98      0.97     13163


Confusion Matrix:
[[   93   231]
 [   49 12790]]


Columns after CountVectorizer 253


How many columns after CountVectorizer? 253

What does that number represent? The size of the vocabulary (i.e., number of distinct API calls across training samples)

What does each row represent? A vectorized version of a malware/goodware API sequence

Do you have out-of-vocabulary in the test set? How many? 253

How does CountVectorizer handle OOV? It simply ignores words not present in the training vocabulary
What classifier did you choose and why? Logistic Regression for a simple baseline with BoW

Final performance? Accuracy, Precision, Recall, F1 (da classification_report)